### PR TITLE
support for conditionals in macros

### PR DESCRIPTION
**Pull request by joehuchette**
Still need to add tests and doc updates.

### CODE REVIEW

**Comment by joehuchette** [comment]
Rebased onto master, this will only work on v0.4 though. Not sure how we feel about syntax that is only supported on a prerelease landing on master?

**Comment by mlubin** [comment]
I don't see an issue as long as it doesn't break anything for 0.3.

File: src/macros.jl [hunk 1]
```diff
@@ -21,15 +21,26 @@ end
...
+    isexpr(c,:ref) || isexpr(c,:typed_vcat) || error("Unrecognized name in construction macro; expected $(string(c)) to be of the form name[...]")
     idxvars = Any[]
     idxsets = Any[]
     idxpairs = IndexPair[]
     # Creating an indexed set of refs
-    cname = c.args[1]
+    cname = shift!(c.args)
     refcall = Expr(:ref,esc(cname))
-    for s in c.args[2:end]
+    if isexpr(c.args[1], :parameters)
```
Comment (mlubin) [inline src/macros.jl]: Does this catch the case where the condition isn't passed as a parameter:
```
julia> dump(:(x[12;3]))
Expr 
  head: Symbol typed_vcat
  args: Array(Any,(3,))
    1: Symbol x
    2: Int64 12
    3: Int64 3
  typ: Any
```

File: test/variable.jl [hunk 2]
```diff
@@ -70,11 +70,13 @@ facts("[variable] repeated elements in index set (issue #199)") do
...
 # Test conditionals in variable definition
-# condmod = Model()
-# @defVar(condmod, x[i=1:10]; iseven(i))
-# @defVar(condmod, y[j=1:10,k=3:2:9]; isodd(j+k) && k <= 8)
-# @test JuMP.dictstring(x, :REPL)   == "x[i], for all i in {1..10} s.t. iseven(i) free"
-# @test JuMP.dictstring(x, :IJulia) == "x_{i} \\quad \\forall i \\in \\{ 1..10 \\} s.t. iseven(i) free"
-# @test JuMP.dictstring(y, :REPL)   == "y[i,j], for all i in {1..10}, j in {3,5..7,9} s.t. isodd(j + k) and k <= 8 free"
-# @test JuMP.dictstring(y, :IJulia) == "y_{i,j} \\quad \\forall i \\in \\{ 1..10 \\}, j \\in \\{ 3,5..7,9 \\} s.t. isodd(j + k) and k <= 8 free"
-# @test string(condmod) == "Min 0\nSubject to \nx[i], for all i in {1..10} s.t. iseven(i) free\ny[i,j], for all i in {1..10}, j in {3,5..7,9} s.t. isodd(j + k) and k <= 8 free\n"
+condmod = Model()
```
Comment (mlubin) [inline test/variable.jl]: We should run these tests only on 0.4, and also add a test that catches any changes in parsing.

**Comment by joehuchette** [comment]
Squashed and rebased, with docs and tests. Think this is about ready to go @IainNZ @mlubin

File: test/macros.jl [hunk 3]
```diff
@@ -295,3 +295,30 @@ facts("[macros] @defExpr") do
 
     @fact_throws @defExpr(blah[i=1:3], x[i,1]^2)
 end
+
+if VERSION >= v"0.4-"
+    facts("[macros] Conditionals in constraint indexing") do
```
Comment (mlubin) [inline test/macros.jl]: Could we add a test for conditions in ``@addNLConstraint``?

**Comment by mlubin** [comment]
Looks pretty good. Note this will break extensions which define their own macros.

File: doc/refexpr.rst [hunk 4]
```diff
@@ -106,6 +106,12 @@ have dependencies on preceding indices, e.g. triangular indexing is allowed::
 
     @addConstraint(m, triconstr[i=1:3,j=2i:2:6], x[i] - y[j] == 1)
 
+A condition can be added following the indices; a semicolon is used to separate index sets from the conditional::
+
+    @addConstraint(m, constr[i=1:5,j=1:5; i+j >= 3], x[i] - y[j] == 1)
+
+Note that only one conditional can be added, although expressions can be built up by using the usual ``&&`` and ``||`` logical operators.
```
Comment (mlubin) [inline doc/refexpr.rst]: s/one conditional/one condition/

File: doc/refvariable.rst [hunk 5]
```diff
@@ -54,11 +54,17 @@ Bounds can depend on variable indices::
...
+    @defVar(m, x[i=1:10,j=i:10] >= 0)
 
 Note the dependency must be on preceding indices, going from left to right. That is,
 ``@defVar(m, x[i=j:10,i=1:10] >= 0)`` is not valid JuMP code.
 
+Conditions can be placed on the index values for which variables are created; the condition follows the statement of the index sets and is separated with a semicolon::
+
+    @defVar(m, x[i=1:10,j=1:10; isodd(i+j)] >= 0)
+
+Note that only one conditional can be added, although expressions can be built up by using the usual ``&&`` and ``||`` logical operators.
```
Comment (mlubin) [inline doc/refvariable.rst]: same here

File: src/macros.jl [hunk 6]
```diff
@@ -429,7 +443,8 @@ macro defExpr(args...)
         error("in @defExpr: needs either one or two arguments.")
     end
 
-    refcall, idxvars, idxsets, idxpairs = buildrefsets(c)
+    crefflag = isa(c,Expr)
```
Comment (mlubin) [inline src/macros.jl]: where is ``crefflag`` used?

File: NEWS.md [hunk 7]
```diff
@@ -1,6 +1,11 @@
 JuMP release notes
 ==================
 
+Unversioned
+-----------
+
+  * Support for conditionals in indexing ``@defVar`` and ``@addConstraint`` constructs, e.g. ``@defVar(m, x[i=1:5,j=1:5; i+j >= 3])
```
Comment (mlubin) [inline NEWS.md]: I'd also say "support for conditions"

File: src/macros.jl (hunk 6 above)
Comment (joehuchette) [inline src/macros.jl]: Guess it isn't here.

**Pull request merged**